title: Post-processing: towards longer motifs selection
date: 2016-09-21
category: notes
tags: alignment, longer motifs, post-processing

## Longer motifs selection

As we illustrated previously, we proceed this issue by focusing on the hot regions generated by the alignments of the selected features. More specifically, we follow the steps listed below.

##### Extraction of the hot regions

Hot regions refer to the most covered patterns by the pre-selected patterns. They can be obtained using *bowtie* and *isolasso*:

```
mkdir origins
bowtie-build origines.fasta origins/genomes
bowtie -a -v 0 --best -r --sam origins/genomes selected_motifs.txt alignedMotifs_0mis.sam
samtools sort -o alignedMotifs_0mis.sorted.sam alignedMotifs_0mis.sam
mkdir isolasso
processsam -g 0 -o isolasso/accepted alignedMotifs_0mis.sorted.sam
```

At this stage, we have got an instance format file: *accepted.instance*. And we can easily extract hot regions by keeping resulting patterns with mean coverage >= threshold.

```
python filter_isolasso.py accepted.instance --o highcov.txt
```


##### Filtering hot regions

We can remove duplicated patterns and patterns that are too short and too long, and keep patterns of length between $low$ and $high$.

```
python rmdup.py highcov.txt low high --o highcov.rmdup.low-high.txt
```

##### Computing feature matrix

We use *bowtie* to compute the feature matrix for this new set of candidate patterns. We can also allow m (<=3) mismatches when creating the matrix.

```
bowtie -a -v m --best -r --sam origins/genomes highcov/highcov.rmdup.low-high.txt highcov/alignedMotifs$m.sam
cd highcov
samtools sort -n alignedMotifs$m.sam | samtools view -o alignedMotifs$m.sorted.sam -
python feature_matrix.py alignedMotifs$m.sorted.sam
```

Until now, we have obtained a feature matrix associated with the set of hot regions generated by pre-selected patterns. We can then apply our method to re-select predictive patterns, which are potentially longer than before. And we anticipate that the prediction error should be smaller.

#### Results

I display here the CV curves for the selection of longer motifs from the novel set of candidate patterns. I have chosen the log-entropy transformation for the feature matrix but without normalization, since we didn't use the whole candidate patterns as before (I hesitate to also remove the normalization in the previous step as we didn't know if y depends on the length of the sequences and then although the prediction error increases...). I have tested different thresholds to filter the hot regions and different values for the mismatch numbers.
And I have found a set of selected motif with near pre-selected performance but with longer motifs. Due to the various choices of filtering and mismatch numbers, the experiment will carry on and I believe that we can obtain better performance.

![postprocessing]({filename}/images/postprocessing/highcov_4_50/cv.png)
![postprocessing]({filename}/images/postprocessing/highcov_4_50/length_dist.png)


## Summarize selected motifs

#### Cluster analysis

One natural approach is to carry out cluster analysis over the resulting selected motifs. We can proceed this by using some kernel PCAs, with kernels related to alignments such as *local alignment kernel* or *substring kernel*. These kernels contain additionally positional information compared to spectrum kernels.

#### Multiple sequence alignment

We can then use an alignment viewer to each of the clusters, in order to visualize and localize the similarities between several motifs and achieve a further subtler and preciser clustering if possible.

#### Results for cluster analysis

I have used 2 kernels to explore clusters within the selected motifs: local alignment kernel and substring kernel. the first contain local positional information and the second allows similarity with gaps. I have also applied a k-medoids algorithm to cluster the projection of the points onto the two first components and displayed the corresponding motifs for the centroids.

![postprocessing]({filename}/images/postprocessing/highcov_4_50/kpca_substring_0.3.png)
![postprocessing]({filename}/images/postprocessing/highcov_4_50/kpca_la.png)
